File: Otros/merf.cpp
#include<iostream>
#include<time.h>
#include<math.h>

using namespace std;

//------------- AYUDA PARA TESTEAR EL CÓDIGO ---------------------------

void random_matrix(float matrix[20][20], int FILAS, int COLUMNAS, int MIN, int MAX) {
        for (short i = 0; i < FILAS; i++) {
                for (short j = 0; j < COLUMNAS; j++) {
                        matrix[i][j] = MIN + rand()%(MAX-MIN);
                }
        }
}

void print_matrix(float matrix[20][20], int FILAS, int COLUMNAS) {
        for (short i = 0; i < FILAS; i++) {
                for (short j = 0; j < COLUMNAS; j++) {
                        cout << matrix[i][j] << "\t";
                }
                cout << "\n";
        }
        cout << "\n";
}

//--------------------------------------------------------------------




//--------- OPERACIONES ELEMENTALES POR FILAS -----------------------

// Multiplicar una fila por un escalar
// i: la fila a modificar
// K: escalar por el que multiplicar
void multiplico(float matrix[20][20], int COL, int i, float K) {
        for(int m = 0; m < COL; m++) {
                matrix[i][m] = K*matrix[i][m];
        }
}

// Multiplicar una fila por una escalar y sumarla a otra
// i1: la fila a objetivo 
// i2: la fila que se va a sumar
// K: escalar por el que multiplicar 
void sumo_multiplico(float matrix[20][20], int COL, int i1, int i2, float K) {
        for(int m = 0; m < COL; m++) {
                matrix[i1][m] += K*matrix[i2][m];
        }
}

// Intercambiar filas
void cambio_fila(float matrix[20][20], int COL, int i1, int i2) {
        float aux;

        for(int m = 0; m < COL; m++) {
                aux = matrix[i1][m];
                matrix[i1][m] = matrix[i2][m];
                matrix[i2][m] = aux;
        }
}

//--------------------------------------------------------------------




// Poncha todas las filas por arriba y por debajo del pivot
// El índice "i" indica que fila no se ponchará (porque es la del pivot)
// El índice "j" sirve para saber por qué coeficiente se deberá multiplicar la fila del pivot
void ponchar(float matrix[20][20], int FIL, int COL, int pivot_i, int pivot_j) {
        for(int i = 0; i < FIL; i++) {
                if(i != pivot_i and matrix[i][pivot_j] != 0) {
                        sumo_multiplico(matrix, COL, i, pivot_i, -matrix[i][pivot_j]);
                }
        }
}


// Elimina los -0 y redondea el resultado a 3 cifras
// Sin esto algunos números no aparecen como 0 sino como números cercanos a 0 (algo tipo -1.14e-8)
// o como -0
void redondear(float matrix[20][20], int FIL, int COL) {
        for(int i = 0; i < FIL; i++) {
                for(int j = 0; j < COL; j++) {
                        matrix[i][j] = round(matrix[i][j] * 1000) / 1000;
                        if(matrix[i][j] == -0) matrix[i][j] = 0;
                }
        }
}

// Busca (por columnas) un número != 0, el cual será el pivot
// Cuando lo encuentra, lo vuelve 1 y poncha las demás filas
// Al final, ordena la matriz
void merf(float matrix[20][20], int FIL, int COL) {
        bool has_pivot;
        int pivot_indx = 0;

        for(int j = 0; j < COL; j++) {
                has_pivot = 0;

                for(int i = pivot_indx; i < FIL; i++) {
                        if(matrix[i][j] != 0.0 and !has_pivot) {
                                has_pivot = 1;

                                if(matrix[i][j] != 1) multiplico(matrix, COL, i, 1.0/matrix[i][j]);
                                
                                cambio_fila(matrix, COL, i, pivot_indx);

                                ponchar(matrix, FIL, COL, pivot_indx, j);
                                pivot_indx++;
                        }
                }
        }

        redondear(matrix, FIL, COL);
}

int main() {
        srand(time(NULL));

        // float matrix[20][20] = {
        //         {0,0,-3,-47},
        //         {-98,-196,-686,-882},
        //         {1,2,7,9}
        // };

        float matrix[20][20] = {
                {0,-3,-47, 3},
                {0,-196,0,-882},
                {0,2,0,9}
        };

        // 8 × 10
        // float matrix[20][20] = {
        //         {-18740,  84935 ,  -14824,  5365  ,  15665  , 59380  , 30079  , 4631   , -50076 , -46021},
        //         {-9097 ,  -70432,  37519 ,  90180 ,  30502  , 10576  , -10719 , -37276 , 76608  , 12702},
        //         {41885 ,  -38439,  55125 ,  12889 ,  -87157 , -43684 , 53783  , 89108  , -40458 , 31854},
        //         {-21713,  40803 ,  16789 ,  -20184,  62520  , 48807  , -60804 , -7401  , -46562 , -94528},
        //         {46578 ,  -39307,  -64960,  -99551,  -49127 , 65543  , -72623 , -43493 , -71733 , -96014},
        //         {-14439,  -13496,  81899 ,  -59313,  15745  , 94742  , -2997  , -30471 , 202    , -27103},
        //         {-98617,  78490 ,  30052 ,  -65476,  -41694 , -7428  , -317   , -86146 , 1524   , -30526},
        //         {35678 ,  64454 ,  30167 ,  70719 ,  64904  , -2607  , -47386 , 8633   , 53900  , 97233}
        // };

        // float matrix[20][20];
        int FIL = 3, COL = 4;

        // random_matrix(matrix, FIL, COL, -10, 10);
        print_matrix(matrix, FIL, COL);

        merf(matrix, FIL, COL);

        print_matrix(matrix, FIL, COL);

        return 0;
}

File: Otros/merf.py
import random as r

PRECISION:int = 3

def p_matrix(matrix:list[list[float]]):
        for fila in matrix:
                print(fila)
        print()

"""
Multiplico una fila por un escalar
i: fila a modificar
K: escalar de la mult.
"""
def type_1_op(matrix:list[list[float]], i:int, K:float):
        for m in range(len(matrix[0])):
                matrix[i][m] = K*matrix[i][m]

"""
Multiplico una fila y la sumo a otra
i1: fila objetivo
i2: fila sumando
K: escalar de la mult.
"""
def type_2_op(matrix:list[list[float]], i1:int, i2:int, K:float):
        for m in range(len(matrix[0])):
                matrix[i1][m] += K*matrix[i2][m]
                matrix[i1][m] = round(matrix[i1][m], PRECISION)

"""
Cambio de filas
"""
def type_3_op(matrix:list[list[float]], i1:int, i2:int):
        aux = matrix[i1]
        matrix[i1] = matrix[i2]
        matrix[i2] = aux

def ponchar(matrix:list[list[float]], pivot_i:int, pivot_j:int):
        for m in range(len(matrix)):
                if m == pivot_i: continue
                type_2_op(matrix, m, pivot_i, -matrix[m][pivot_j])

def round_values(matrix:list[list[float]]):
        for i in range(len(matrix)):
                for j in range(len(matrix[0])):
                        matrix[i][j] = round(matrix[i][j], PRECISION)

def sort(matrix:list[list[float]]):
        index:int = 0

        for j in range(len(matrix[0])):
                for i in range(len(matrix)):
                        if(matrix[i][j] != 0 and index < len(matrix)):
                                type_3_op(matrix, index, i)
                                index += 1
                                break

def merf(matrix:list[list[float]]):
        has_pivot:bool

        for i in range(len(matrix)):
                has_pivot = False

                for j in range(len(matrix[0])):
                        if matrix[i][j] != 0 and not has_pivot:
                                type_1_op(matrix, i, 1/matrix[i][j])
                                has_pivot = True
                                ponchar(matrix, i, j)
        
        round_values(matrix)
        sort(matrix)

def main():
        m:int = r.randint(2, 5)
        n:int = r.randint(2, 5)

        print(f"Dimensiones: {m}×{n}\n")

        matrix:list[list[float]] = [[r.randint(-10, 10) for j in range(n)] for i in range(m)]
        
        p_matrix(matrix)
        merf(matrix)
        p_matrix(matrix)
        

if __name__ == '__main__':
        main()